test/performance/scale_from_zero_test.go

```diff
@@ -0,0 +1,192 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package performance
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"testing"
+	"time"
+
+	pkgTest "github.com/knative/pkg/test"
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/test-infra/shared/testgrid"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/api/extensions/v1beta1"
+
+	"github.com/knative/serving/test"
+)
+
+const (
+	serviceName                      = "perftest-scalefromzero"
+	ScaleFromZeroAvgTestGridProperty = "perf_ScaleFromZero_Average"
+	ScaleFromZeroTestGridClass       = "ScaleFromZero"
+)
+
+const (
+	helloWorldExpectedOutput = "Hello World!"
+)
+
+type stats struct {
+	avg time.Duration
+}
+
+func isDeploymentScaledToZero() func(d *v1beta1.Deployment) (bool, error) {
+	return func(d *v1beta1.Deployment) (bool, error) {
+		return d.Status.ReadyReplicas == 0, nil
+	}
+}
+
+func runScaleFromZero(clients *test.Clients, logger *logging.BaseLogger, ro *test.ResourceObjects) (time.Duration, error) {
+	deploymentName := ro.Config.Status.LatestCreatedRevisionName + "-deployment"
+
+	domain := ro.Route.Status.Domain
+	logger.Info("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		isDeploymentScaledToZero(),
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		return 0, fmt.Errorf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	start := time.Now()
+	logger.Info("Waiting for endpoint to serve request")
+	if _, err := pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		logger,
+		domain,
+		pkgTest.Retrying(pkgTest.MatchesBody(helloWorldExpectedOutput), http.StatusNotFound),
+		"HelloWorldServesText",
+		test.ServingFlags.ResolvableDomain); err != nil {
+		return 0, fmt.Errorf("The endpoint for Route %s at domain %s didn't serve the expected text \"%s\": %v", ro.Route.Name, domain, helloWorldExpectedOutput, err)
+	}
+
+	logger.Info("Request completed")
+	return time.Now().Sub(start), nil
+}
+
+func parallelScaleFromZero(t *testing.T, logger *logging.BaseLogger, count int) ([]time.Duration, error) {
+	ctx := context.TODO()
+	pc, err := Setup(ctx, logger, false)
+	if err != nil {
+		t.Fatalf("Failed to setup clients: %v", err)
+	}
+
+	testNames := make([]*test.ResourceNames, count)
+	durations := make([]time.Duration, count)
+
+	// Initialize our service names
+	for i := 0; i < count; i++ {
+		testNames[i] = &test.ResourceNames{
+			Service: test.AppendRandomString(serviceName, logger),
```

Comment:
do you want to include index as part of the random name for easier traceability in logs?